File: 046_permutations/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> results = new ArrayList<>();
        dfs(results, nums, new ArrayList<>(), new int[nums.length]);

        return results;
    }

    public void dfs(List<List<Integer>> results, int[] nums, ArrayList<Integer> tmp, int[] visited) {
        if (tmp.size() == nums.length) {
            results.add(new ArrayList<>(tmp));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited[i] == 1) continue;
            tmp.add(nums[i]);
            visited[i] = 1;
            dfs(results, nums, tmp, visited);
            visited[i] = 0;
            tmp.remove(tmp.size() - 1);
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1, 2, 3};
        Solution solution = new Solution();
        List<List<Integer>> results = solution.permute(nums);
        for (List<Integer> result : results) {
            for (int ele : result) {
                System.out.print(ele + " ");
            }
            System.out.println();
        }
    }
}

File: 046_permutations/Solution.go
package main

import "fmt"

func permute(nums []int) [][]int {
    var results [][]int
    dfs(&results, nums, []int{})

    return results
}

func dfs(results *[][]int, nums, tmp []int) {
    if len(nums) == 0 {
        *results = append(*results, tmp)
        return
    }

    for i := 0; i < len(nums); i++ {
        dfs(results, combinedArr(nums[:i], nums[i+1:]), combinedArr(tmp, []int{nums[i]}))
    }
}

func combinedArr(arr1, arr2 []int) []int {
    var result []int
    result = append(result, arr1...)
    result = append(result, arr2...)

    return result
}

func main() {
    nums := []int{1, 2, 3}
    results := permute(nums)
    for _, result := range results {
        fmt.Println(result)
    }
}

